Create `@view` macro for creating SubArrays via indexing.

simonbyrne opened the pull request:
Allows writing things like

``` julia
@sub A[1,2:end]
```

In order to replicate current indexing behaviour, this actually calls `slice`.

I'm fairly confident this replicates the `end` replacing behaviour in https://github.com/JuliaLang/julia/blob/c43b5e5e716cf42cf906a4bc16be06846aaf7707/src/julia-syntax.scm#L93-L123

This was discussed in an issue at some point, but I forgot which one.

simonbyrne committed:
Create `@sub` macro for creating SubArrays via indexing.

tkelman commented on base/subarray.jl:
```diff
@@ -352,3 +352,76 @@ function parentdims(s::SubArray)
+end
+
+"""
+    @sub A[...]
```
should be added to rst docs somewhere if exported

simonbyrne replied:
How do I do that? I couldn't figure it out.

tkelman replied:
It's somewhat documented in CONTRIBUTING.md I think. The gist of it is pick a place for it in the rst, add just the signature, then `make docs` should fill in the rest.

simonbyrne replied:
Great, thanks.

simonbyrne replied:
Done.

tkelman commented:
Doing this as a macro means it can't support things like `A[1:end-1]`, can it?

mbauman commented:
Nice! Some previous discussion is buried in the Arraypocalype issue and starts here: https://github.com/JuliaLang/julia/issues/13157#issuecomment-176647475

One question is if the macro should recurse through all non-assignment refs or not.  I think I'd prefer the recursive behavior, allowing for `@sub begin #= algorithm #= end` to easily use views everywhere.

simonbyrne commented:
Recursive behaviour is more complicated, and we would need to figure out some possible ambiguities like:
```julia
@sub A[B[2:end]]
```
Should that create subarrays of both `A` and `B`?

I figured this was a reasonable start, and recursive behaviour could be added later.

simonbyrne commented:
> hm what was it about end that needed to be handled by lowering then?

I'm not sure, but this might require input from someone who knows more about the internals.

tkelman commented:
I think I was just misreading the caveats about the macro option from the long discussion, it was warned that you'd have to reimplement the `end` lowering behavior in the macro which is exactly what you've done here. More complicated than just replacing `getindex` with `slice`, but not too many lines of code overall. One step closer to having a pure-Julia implementation of all lowering?

tkelman commented on base/subarray.jl:
```diff
@@ -352,3 +352,76 @@ function parentdims(s::SubArray)
+"""
+    @sub A[inds...]
+
+Creates `SubArray` from an indexing expression. This can only be applied directly to a
```
"Creates a `SubArray`"

simonbyrne replied:
good catch.

cstjean commented:
#15750 is relevant: *use first/last as endpoints for Range indexing*

timholy commented:
I'm nervous about #15750 because I worry it can't handle all situations. This is certainly safer.

Hopefully in the not too distant future we'll also have to generalize this for #16260, and decide whether to add support for `begin`.

timholy commented:
> hm what was it about end that needed to be handled by lowering then?

The point has always been that we couldn't handle `sub(A, 2:end)` because the lowering looks for the containing `Expr(:ref, ...)`, and when you use `end` as an argument inside a function there isn't such a containing expression. In this case, the macro gets to work with such an expression, so all is well.

I should also point out that Jeff has made a related proposal, adding `A@[2:end]` to create a view. On balance, though, I think I prefer `@sub`, mainly because someday I will probably want to add `@index` and `@stored` too (see [ArrayIteration.jl](https://github.com/timholy/ArrayIteration.jl)).

simonbyrne commented:
> I should also point out that Jeff has made a related proposal, adding `A@[2:end]` to create a view. On balance, though, I think I prefer `@sub`, mainly because someday I will probably want to add `@index` and `@stored` too (see ArrayIteration.jl).

The thought just occurred to me that we could use `A@sub[2:end]`, which would simply be syntactic sugar for `@sub(A[...])`? This would then encompass your other examples.

tkelman commented:
There's some sentiment in #13157 that this might be better named `@view`, ArrayViews use can be qualified in packages.

simonbyrne commented:
I originally proposed it as `@view`, but I thought `@sub` might be better since, (a) it's 1 character shorter, and (b) it calls `sub` (okay, it actually calls `slice`, but I presume we're changing that?).

simonbyrne commented:
Also worth providing a link to the original inspiration https://github.com/JuliaLang/ArrayViews.jl/pull/14

simonbyrne commented:
I've renamed to `@view` to mirror #16972. I've also changed it so that it expands to `true && view(X, inds...)`, to prevent people redefining view by doing `@view(X[inds...]) = ...`

tkelman commented:
Merge?
